src/modules/Game.class.js

```diff
@@ -1,68 +1,301 @@
+            newRow.splice(i - 1, 1);
+            merged.splice(i - 1, 1);
+            merged[i - 1] = true;
+          }
+        }
+
+        while (newRow.length < size) {
+          newRow.unshift(0);
+        }
+
+        for (let col = 0; col < size; col++) {
+          if (state[row][col] !== newRow[col]) {
+            state[row][col] = newRow[col];
+            moved = true;
+          }
+        }
+      }
+
+      this.handleMove(moved);
+    }
   }
 
-  moveLeft() {}
-  moveRight() {}
-  moveUp() {}
-  moveDown() {}
-
-  /**
-   * @returns {number}
-   */
-  getScore() {}
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  moveUp() {
+    if (this.getStatus() === Game.STATUS.PLAYING) {
+      const size = this.state.length;
+      const state = this.getState();
+      let moved = false;
+
+      for (let col = 0; col < size; col++) {
+        const newColumn = [];
+
+        for (let row = 0; row <= size - 1; row++) {
+          if (state[row][col] !== 0) {
+            newColumn.push(state[row][col]);
+          }
+        }
+
+        for (let i = 0; i < newColumn.length; i++) {
+          if (newColumn[i] === newColumn[i + 1]) {
+            newColumn[i] *= 2;
+            this.score += newColumn[i];
+            newColumn.splice(i + 1, 1);
+          }
+        }
+
+        while (newColumn.length < size) {
+          newColumn.push(0);
+        }
+
+        for (let row = 0; row < size; row++) {
+          if (state[row][col] !== newColumn[row]) {
+            state[row][col] = newColumn[row];
+            moved = true;
+          }
+        }
+      }
+
+      this.handleMove(moved);
+    }
+  }
+
+  moveDown() {
+    if (this.getStatus() === Game.STATUS.PLAYING) {
+      const size = this.state.length;
+      const state = this.getState();
+      let moved = false;
+
+      for (let col = 0; col < size; col++) {
+        const newColumn = [];
+
+        for (let row = 0; row <= size - 1; row++) {
+          if (state[row][col] !== 0) {
+            newColumn.push(state[row][col]);
+          }
+        }
+
+        const merged = Array(newColumn.length).fill(false);
+
+        for (let i = newColumn.length - 1; i > 0; i--) {
+          if (
+            newColumn[i] === newColumn[i - 1] &&
+            !merged[i] &&
+            !merged[i - 1]
+          ) {
+            newColumn[i] *= 2;
+            this.score += newColumn[i];
+            newColumn.splice(i - 1, 1);
+            merged.splice(i - 1, 1);
+            merged[i - 1] = true;
```

Comment:
The logic in 'moveDown' for merging tiles is correct, but it is similar to the logic in the other move methods. Consider refactoring to reduce code duplication and centralize the merging logic into a single method.